enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,580 @@
+
+>An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
+
+OLM could  handle the scenario described above had the operator had the ability to communicate that it should not be upgraded as it ran the critical process.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate a condition that influences OLM behavior, such as the "Upgrade Readiness" condition defined above.
+- Preserve existing behavior if the operator does not take advantage of this feature.
+
+### Non-Goals
+
+- Provide operators with the means to cancel an upgrade that is in process.
+- OLM cannot guarantee that some outside force will not interrupt the services of the operator and will not provide the operator with the ability to communicate that it should not be "Interrupted".
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate conditions to OLM that it cannot infer otherwise.
+
+#### Story 2
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade my operator until I tell it otherwise.
+
+## Design Details
+
+OLM will introduce the `Probe` CustomResourceDefinition which will communicate an operator's state through the use of [conditions](https://github.com/kubernetes/enhancements/blob/3aa92e20bdfa2e60e3cb1a2b92cdca61847d7ad2/keps/sig-api-machinery/1623-standardize-conditions/README.md#kep-1623-standardize-conditions). For those unfamiliar with conditions, they are used by many Kubernetes Resources to provide users with additional insight into the state of the resource and typically adhere to the following format:
+
+| Field | Description |
+| ----- | ----------- |
+| Type | Type of condition in CamelCase or in foo.example.com/CamelCase. |
+| Status | Status of the condition, one of True, False, Unknown. |
+| ObservedGeneration | If set, this represents the .metadata.generation that the condition was set based upon. |
+| LastTransitionTime | Last time the condition transitioned from one status to another. |
+| Reason | The reason for the condition's last transition in CamelCase. |
+| Message | A human readable message indicating details about the transition. |
+
+OLM will define a set of "OLM Supported Conditions" which will influence the behavior of OLM when present. The list of "OLM Supported Conditions" can be seen in the table below:
+
+| Condition Type | Effect |
+| -------------- | ------ |
+| Upgradeable | When `false`, OLM will not upgrade the operator.|
```

Comment:
In OpenShift, we pivoted `Upgradeable` to only block minor+ bumps, and to no longer block patch bumps ([rhbz#1797624][1]).  That allows you to gracefully recover from:

> I released a broken patch which sets `Upgradeable=False` more aggressively than I meant, and now I want to left folks perform patch updates out of that version without having to jump through whatever hoops the overly-strict version required to clear `Upgradeable=False`.

However, it also relies on two assumptions:

* Operator workloads are always interruptible for patch bumps.  Which you definitely want, but this enhancement seems maybe not ready to commit to (e.g. [here][2]).
* Patch bumps never make any changes large enough that you'd actually _want_ to block them.  Which you also want, but I have no idea how well folks SemVer their operators, if they SemVer them at all...

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1797624
[2]: https://github.com/operator-framework/enhancements/pull/23#discussion_r441723661